### ISSUE TITLE
Controller spewing out new and new pods when predicate fails

### ISSUE THREAD

**Issue by CallMeFoxie** [kubernetes/kubernetes#66381]
/kind bug

**What happened**:
- Scenario: 2 nodes in a cluster.
- Deployment with hostPort
- I request more instances of the given pod
- Schedules but fails to start because overlapping hostports on the node(s)

**What you expected to happen**:
Have one pod pending

**How to reproduce it (as minimally and precisely as possible)**:
- have let's say 2 nodes
- deploy this (or with 1 replica and scale up):
```
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app-name: test
  name: test
  namespace: default
spec:
  progressDeadlineSeconds: 600
  replicas: 3
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app-name: test
  strategy:
    rollingUpdate:
      maxSurge: 1
      maxUnavailable: 1
    type: RollingUpdate
  template:
    metadata:
      creationTimestamp: null
      labels:
        app-name: test
      name: test
    spec:
      containers:
      - image: <any image>
        imagePullPolicy: IfNotPresent
        name: memcache
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        ports:
        - containerPort: 5001
          hostPort: 8002
          name: nginx
          protocol: TCP
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
```
it will keep creating new pods which will fail into PodFitsHostPorts condition and immediately create yet another and more and more pods that fail.

```
  Type     Reason            Age   From                               Message
  ----     ------            ----  ----                               -------
  Warning  FailedScheduling  8m    default-scheduler                  0/2 nodes are available: 1 node(s) didn't have free ports for the requested pod ports, 1 node(s) didn't match node selector.
  Normal   Scheduled         8m    default-scheduler                  Successfully assigned img/test-whatever to somenode
  Warning  PodFitsHostPorts  8m    kubelet, somenode  Predicate PodFitsHostPorts failed
```

**Environment**:
- Kubernetes version (use `kubectl version`): 1.11.1
- Cloud provider or hardware configuration: on-prem
- OS (e.g. from /etc/os-release): debian stretch
- Kernel (e.g. `uname -a`):debian 4.9
- Install tools: built from source

Ashley

**Comment by CallMeFoxie** [comment]
Ok another observation: There needs to be two different deployments fighting over that port it seems, just one does not trigger this with high chance. When there's two fighting it goes much faster with higher chance.

**Comment by ravisantoshgudimetla** [comment]
What is happening to old pod? Is it getting killed?

**Comment by ravisantoshgudimetla** [comment]
So, if kubelet is not admitting, then scheduler should have seen it as well as they share the same code for predicates. If you could provide kubelet logs, it could further debugging.

**Comment by k82cn** [comment]
> There needs to be two different deployments fighting over that port it seems

What the yaml file of the other Deployment? It's better to use different selector of each Deployments.

```yaml
.....

selector:
    matchLabels:
      app-name: test

....

      labels:
        app-name: test
```

**Comment by CallMeFoxie** [comment]
Ah yeah, of course they have different selectors etc, they were even in different namespaces when I observed it and went to diagnose the bug :) Just need to have the same hostport exported.

Somehow I am unable to replicate the issue today (yay friday) and journal has already removed yesterday's logs :( Will keep on trying.

**Comment by CallMeFoxie** [comment]
Actually the PodFitsHostPort pods have Failed state, not terminated/evicted.

**Comment by CallMeFoxie** [comment]
OK one precondition I forgot to mention and forgot about today: I've spammed `kube-dns` pods beyond what the cluster can take (`requests` and `limits`) and not all got scheduled. However the kube-dns has got high priority whilst the other two deployments have normal default priority of zero.

I see nothing in logs of kubelet BUT kubelet is boiling 100% of the CPU.

```
Jul 20 10:59:22 hostname kubelet[36682]: I0720 10:59:22.200230   36682 eviction_manager.go:317] eviction manager: no resources are starved
Jul 20 10:59:22 hostname kubelet[36682]: I0720 10:59:22.258093   36682 config.go:303] Setting pods for source api
Jul 20 10:59:22 hostname kubelet[36682]: I0720 10:59:22.261478   36682 config.go:414] Receiving a new pod "img-searchapi-859f47c74f-mm7gb_img(30516062-8bfb-11e8-adc7-5cf3fc4a7c5c)"
Jul 20 10:59:22 hostname kubelet[36682]: I0720 10:59:22.274610   36682 kubelet.go:1837] SyncLoop (ADD, "api"): "img-searchapi-859f47c74f-mm7gb_img(30516062-8bfb-11e8-adc7-5cf3fc4a7c5c)"
Jul 20 10:59:22 hostname kubelet[36682]: I0720 10:59:22.276926   36682 predicate.go:133] Predicate failed on Pod: img-searchapi-859f47c74f-mm7gb_img(30516062-8bfb-11e8-adc7-5cf3fc4a7c5c), for reason: Predicate PodFitsHostPorts failed
Jul 20 10:59:22 hostname kubelet[36682]: I0720 10:59:22.277059   36682 server.go:460] Event(v1.ObjectReference{Kind:"Pod", Namespace:"img", Name:"img-searchapi-859f47c74f-mm7gb", UID:"30516062-8bfb-11e8-adc7-5cf3fc4a7c5c", APIVersion:"v1", ResourceVersion:"1215000761", FieldPath:""}): type: 'Warning' reason: 'PodFitsHostPorts' Predicate PodFitsHostPorts failed
Jul 20 10:59:22 hostname kubelet[36682]: I0720 10:59:22.282292   36682 config.go:303] Setting pods for source api
```
on kubelet

**Comment by CallMeFoxie** [comment]
I wonder if it is some race condition with eviction on equal pods of the `kube-dns` and scheduler doing ... -- hmm ok I have no idea here, anyone? :)

**Comment by CallMeFoxie** [comment]
This is the kube-dns deployment:
```
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "11"
  creationTimestamp: 2017-02-06T08:54:19Z
  generation: 28
  labels:
    k8s-app: kube-dns
  name: kube-dns
  namespace: kube-system
  resourceVersion: "1214999607"
  selfLink: /apis/extensions/v1beta1/namespaces/kube-system/deployments/kube-dns
  uid: d8ae25e2-ec49-11e6-a43c-0894ef239b21
spec:
  progressDeadlineSeconds: 600
  replicas: 40
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      k8s-app: kube-dns
  strategy:
    rollingUpdate:
      maxSurge: 30%
      maxUnavailable: 30%
    type: RollingUpdate
  template:
    metadata:
      creationTimestamp: null
      labels:
        k8s-app: kube-dns
      name: kube-dns
    spec:
      containers:
      - args:
        - --domain=cluster.local.
        - --dns-port=53
        - --kube-master-url=https://172.31.255.1:443
        - --v=2
        image: doc.ker/admins5/kubedns-amd64:1.14.1
        imagePullPolicy: IfNotPresent
        livenessProbe:
          failureThreshold: 5
          httpGet:
            path: /healthz
            port: 8080
            scheme: HTTP
          initialDelaySeconds: 60
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 5
        name: kubedns
        ports:
        - containerPort: 53
          name: dns
          protocol: UDP
        - containerPort: 53
          name: dns-tcp
          protocol: TCP
        readinessProbe:
          failureThreshold: 3
          httpGet:
            path: /readiness
            port: 8081
            scheme: HTTP
          initialDelaySeconds: 20
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 5
        resources:
          limits:
            memory: 170Mi
          requests:
            cpu: "2"
            memory: 70Mi
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /srv/kubernetes
          name: srvkube
          readOnly: true
      - args:
        - -cmd=nslookup kubernetes.default.svc.cluster.local 127.0.0.1 >/dev/null
        - -port=8080
        - -quiet
        image: doc.ker/admins5/exechealthz-amd64:1.2-fix
        imagePullPolicy: IfNotPresent
        name: healthz
        ports:
        - containerPort: 8080
          protocol: TCP
        resources:
          limits:
            memory: 50Mi
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      dnsPolicy: Default
      nodeSelector:
        isapiserver: "true"
**      priorityClassName: system-cluster-critical**
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      serviceAccount: kube-dns
      serviceAccountName: kube-dns
      terminationGracePeriodSeconds: 30
      volumes:
      - hostPath:
          path: /www/kubernetes/dns
          type: ""
        name: srvkube
      - hostPath:
          path: /etc/ssl
          type: ""
        name: etcssl
```

however the nodes are completely full (both have +- the same HW and capacity and load):
```
Allocated resources:
  (Total limits may be over 100 percent, i.e., overcommitted.)
  CPU Requests  CPU Limits  Memory Requests  Memory Limits
  ------------  ----------  ---------------  -------------
  40 (100%)     0 (0%)      2400Mi (3%)      4400Mi (6%)
```
which prevents the remaining instances of kube-dns be scheduled just as the test pods and they're somehow fighting over.
I have no idea how the `PodFitsHostPorts` clicks in though.

**Comment by CallMeFoxie** [comment]
Still happening with 1.11.2.

Seeing this on --v=6 on scheduler
```
Aug 10 12:56:24 hostname hyperkube[18607]: I0810 12:56:24.566981   18607 factory.go:1181] About to try and schedule pod test-74d79cc65b-rt26b
Aug 10 12:56:24 hostname hyperkube[18607]: I0810 12:56:24.566997   18607 scheduler.go:447] Attempting to schedule pod: default/test-74d79cc65b-rt26b
Aug 10 12:56:24 hostname hyperkube[18607]: I0810 12:56:24.567411   18607 scheduler.go:194] Failed to schedule pod: default/test-74d79cc65b-rt26b
Aug 10 12:56:24 hostname hyperkube[18607]: I0810 12:56:24.567434   18607 factory.go:1303] Unable to schedule default test-74d79cc65b-rt26b: no fit: 0/2 nodes are available: 1 node(s) didn't have free ports for the requested pod ports, 2 Insufficient cpu.; waiting
Aug 10 12:56:24 hostname hyperkube[18607]: I0810 12:56:24.567492   18607 factory.go:1416] Updating pod condition for default/test-74d79cc65b-rt26b to (PodScheduled==False)
Aug 10 12:56:24 hostname hyperkube[18607]: I0810 12:56:24.607456   18607 round_trippers.go:405] PUT http://localhost:8080/api/v1/namespaces/default/pods/test-74d79cc65b-rt26b/status 200 OK in 8 milliseconds
Aug 10 12:56:2 hostname hyperkube[18607]: I0810 12:56:24.619258   18607 request.go:485] Throttling request took 51.349569ms, request: GET:http://localhost:8080/api/v1/namespaces/default/pods/test-74d79cc65b-rt26b
Aug 10 12:56:24 hostname hyperkube[18607]: I0810 12:56:24.625118   18607 round_trippers.go:405] GET http://localhost:8080/api/v1/namespaces/default/pods/test-74d79cc65b-rt26b 200 OK in 5 milliseconds
Aug 10 12:56:24 hostname hyperkube[18607]: I0810 12:56:24.643118   18607 round_trippers.go:405] GET http://localhost:8080/api/v1/namespaces/default/pods/test-74d79cc65b-rt26b 200 OK in 4 milliseconds
```
repeated forever and ever, no idea where is the PodFitsHostPorts appearing from, the `kubelet` part of it seems wrong as I see nothing in the logs of kubelets about this at all.

Is it some default/leftover?

I think the main issue is that it is spewing more and more pods and I am unable to track down why - controller is only logging "created new pod", just not why even on `--v=6` loglevel. I presume it is because the previous one failed scheduling, but it should not create new pods?

**Comment by CallMeFoxie** [comment]
Ok https://github.com/kubernetes/kubernetes/issues/67453 has fixed similar issue, however adding it does not fix this. I wonder if it is a similar code somewhere else? 

Also it seems that the pod gets assigned to some node, which does not have this port available, kubelet reports PodFitsHostPort status and controller creates a new pod etc etc.

**Comment by Huang-Wei** [comment]
@CallMeFoxie the message "PodFitsHostPort" in kubelet is actually misleading - it doesn't share the same predicates logic of scheduler, instead it reads the *real* host occupation info from its running container interface. The status showed "Failed" also proves that - a pod is scheduled to a host, but kubelet rejects it as there is already an application using that port.

Then you might ask: why scheduler schedules pod to that host? The answer is that unlike kubelet, scheduler doesn't use *real* host occupation info. The logic in scheduler is simply like this: for a to-schedule pod, if it's "claim"ing for a specific hostPort, loop over all nodes and pods, find out all nodes which doesn't have pods "claim"ing that port.

Note that "claim" here means in k8s API perspective: the pod should be specifying `hostNetwork` and have `containerPort/hostPort` defined.

Back to your case, I _guess_ possible failure reasons are:

1.  you're not specifying `hostNetwork` and(or) `containerPort/hostPort` in workload manifest
2. there is an app which is out of k8s context using that port

**Comment by CallMeFoxie** [comment]
Nope, I am not using hostNetwork on purpose and hostPort is specified on purpose - otherwise it wouldn't be bound on the host. (and I made sure there is no outgoing connection/.. anything bound on that port - I can 100% redo the test even after rebooting the whole cluster etc).

I added some debug info messages and when it tries to schedule some overrunning kube-dns pods it shows that the used host ports contain the port already allocated by the other deployment, however it seems that during scheduling of these pods it is either empty or the check is completely missing (need to rebuild it with extra debug info). Will post tomorrow.

**Comment by Huang-Wei** [comment]
> Ok #67453 has fixed similar issue, however adding it does not fix this. I wonder if it is a similar code somewhere else?

@CallMeFoxie when you patch #67453, in addition to kube-scheduler, is kubelet also rebuilt and restarted? (I asked this as this problem seems to be a kubelet issue - from the error, it seems kubelet hasn't went to the step to check *realtime* hostPort info yet. Rather failing on reusing scheduler predicates logic to do sort of a "precheck")

**Comment by Huang-Wei** [comment]
And I have no luck to reproduce using latest master code. Here are the steps:

```
# k version --short
Client Version: v1.13.0-alpha.0.425+c0be4cc09c5a68
Server Version: v1.13.0-alpha.0.425+c0be4cc09c5a68
```

- create a 10 node cluster
- prepare two yamls - n1.yaml, n2.yaml
    ```yaml
    $ cat n1.yaml
    apiVersion: apps/v1
    kind: Deployment
    metadata:
      name: nginx1
    spec:
      replicas: 6
      selector:
        matchLabels:
          app: nginx1
      template:
        metadata:
          labels:
            app: nginx1
        spec:
          containers:
          - name: nginx1
            image: nginx
            ports:
            - containerPort: 80
              hostPort: 9090
    $ sed 's/nginx1/nginx2/' n1.yaml > n2.yaml
    ```
- create n1.yaml and n2.yaml
- as expected, 2 pods (6*2-10) are in Pending status
- keep deleting all pods (`k delete po --all`) and observe
- I tried several times but everytime there are always 2 (random) pods in Pending status - unlike what you said some pods are being recreated

**Comment by CallMeFoxie** [comment]
Seems that building from a master branch and restarting everything (apiserver/controller/scheduler and most importantly all the kubelets) indeed did the trick, will see if it is fixed in 1.11.3 branch or if it is something else from master branch. :) Thanks @Huang-Wei !

**Comment by Huang-Wei** [comment]
@CallMeFoxie Great to hear that!

**Comment by CallMeFoxie** [comment]
Yep seems to be fixed with 1.11.3 release @Huang-Wei ! Thanks :)
